### DISCUSSION TITLE
Setting the RTC of T4.x

### DISCUSSION THREAD

**Discussion by luni64** [luni64/TeensySharp#5] [General]
As shown by @mecparts here https://github.com/Koromix/tytools/issues/41#issuecomment-748415067 it is possible to set the RTC time by adding data to the reset command of the bootloader. I implemented this in TeensySharp (development branch) and generated a quick example app.  Works perfectly! Thanks  @mecparts

Here a video showing it in action:   (first seconds don't play correctly...)

https://user-images.githubusercontent.com/12611497/102710919-41c76600-42b6-11eb-91fd-d9d5e849a734.mp4

Here the Teensy sketch I used: 
```c++
#include "Arduino.h"
#include <ctime>

void setup()
{
  pinMode(LED_BUILTIN,OUTPUT);
}

void loop()
{
    time_t t = rtc_get();
    SerialUSB1.print(std::ctime(&t));

    digitalToggle(LED_BUILTIN);
    delay(500);
}
```
And here the addition to the reset procedure: 
```c#
  var report = new byte[reportSize];  
  report[0] = 0x00;                                       // hid report nr
  BitConverter.GetBytes(0x00FF_FFFF).CopyTo(report, 1);   // block addr (0x00FF_FFFF -> reset)
  if (time != null)
  {
       DateTime epoch = DateTime.ParseExact("1970-01-01_00:00:00", "yyyy-MM-dd_HH:mm:ss", CultureInfo.InvariantCulture);
       uint now = (uint)(time.Value - epoch).TotalSeconds;

       BitConverter.GetBytes(0x89C2_31B7).CopyTo(report, 65);  // magic number?
       BitConverter.GetBytes(now << 15).CopyTo(report, 69);    // SNVS_LPSTRCLR
       BitConverter.GetBytes(now >> 17).CopyTo(report, 73);    // SNVS_LPSRTCMR
  }
  result = hidWriteReport(hidHandle, report) ? ErrorCode.OK : ErrorCode.ResetError;
```

If somebody want to try, here the binaries: [SetTime.zip](https://github.com/luni64/TeensySharp/files/5720790/SetTime.zip)

**Comment by Defragster** [comment]
That worked here!  Took TyComm offline, ran EXE and on restart the sketch edited to print time worked.
ReUploaded that Sketch with TyComm and the Time is there again/still.

" ...
 1 dirs with 4 files of Size 5557590 Bytes

Sun Dec 20 02:43:04 2020"

**Comment by luni64** [comment]
Something went wrong with the last part of your messge?

I forgot to mention that I used SerialUSB1 for the time output so that I can reset the teensy while it communicates on the second port. Of course this only works if you switch the  ty-sermon (main port) off in TyComm

**Comment by Defragster** [comment]
No that was the message.  Just showing printed Date and Time from current Sketch LittleFS disk in beta - 128GB NAND Flash.

I saw it was USB1 - and changed to print to Serial. And yes turning off TyComm Serial is needed.
